lib/tree.rb

```diff
@@ -16,47 +16,141 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def add(key, value=nil)
+    new_node = TreeNode.new(key, value)
+    if @root == nil 
+      @root = new_node 
+    else
+      add_helper(@root, new_node)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    return new_node if current == nil
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+
+    return current
+  end
+
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
+  end
+
+  def find_helper(current, key)
+    return nil if current == nil
+
+    if current.key == key
+      return current.value
+    elsif key <= current.key
+      return find_helper(current.left, key)
+    else key > current.key
+      return find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    keys_values = []
+    current = @root
+    return inorder_helper(@root, keys_values)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, keys_values)
+    return keys_values if current == nil
+
+    inorder_helper(current.left, keys_values)
+    keys_values.push({:key => current.key, :value => current.value})
+    inorder_helper(current.right, keys_values)
+    return keys_values
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    keys_values = []
+    current = @root
+    return preorder_helper(@root, keys_values)
+  end
+
+  def preorder_helper(current, keys_values)
+    return keys_values if current == nil
+
+    keys_values.push({:key => current.key, :value => current.value})
+    preorder_helper(current.left, keys_values)
+    preorder_helper(current.right, keys_values)
+
+    return keys_values
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    keys_values = []
+    current = @root
+    return postorder_helper(@root, keys_values)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current, keys_values)
+    return keys_values if current == nil
+
+    postorder_helper(current.left, keys_values)
+    postorder_helper(current.right, keys_values)
+    keys_values.push({:key => current.key, :value => current.value})
+  
+    return keys_values
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
👍 The space complexity is only O(n) if the tree is unbalanced and O(log n) if it's balanced.